README.md
<h1 align="center">
  <img src="https://raw.githubusercontent.com/gregberge/xstyled/master/resources/xstyled-logo.jpg" alt="xstyled" title="xstyled" width="300">
</h1>
<p align="center" style="font-size: 1.2rem;">A utility-first CSS-in-JS framework built for React.</p>

[![License](https://img.shields.io/npm/l/@agriffis/xstyled-styled-components.svg)](https://github.com/gregberge/xstyled/blob/master/LICENSE)
[![npm package](https://img.shields.io/npm/v/@agriffis/xstyled-styled-components/latest.svg)](https://www.npmjs.com/package/@agriffis/xstyled-styled-components)
[![npm downloads](https://img.shields.io/npm/dm/@agriffis/xstyled-styled-components.svg)](https://www.npmjs.com/package/@agriffis/xstyled-styled-components)
[![CircleCI](https://circleci.com/gh/gregberge/xstyled.svg?style=svg)](https://circleci.com/gh/gregberge/xstyled)
[![codecov](https://codecov.io/gh/gregberge/xstyled/branch/master/graph/badge.svg)](https://codecov.io/gh/gregberge/xstyled)
![Code style](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)

```bash
npm install @agriffis/xstyled-styled-components styled-components
```

## [Docs](https://xstyled.dev)

**See the documentation at [xstyled.dev](https://xstyled.dev)** for more information about using xstyled!

Quicklinks to some of the most-visited pages:

- [**Getting started**](https://xstyled.dev/docs/installation/)
- [Motivation](https://xstyled.dev/docs/introduction/#story)

## Example

```js
import { x } from '@agriffis/xstyled-styled-components'

function Example() {
  return (
    <x.div p={{ _: 3, md: 6 }} bg="white" display="flex" spaceX={4}>
      <x.div flexShrink={0}>
        <x.img h={12} w={12} src="/img/logo.svg" alt="xstyled Logo" />
      </x.div>
      <x.div>
        <x.h4
          fontSize={{ _: 'md', lg: 'xl' }}
          fontWeight="medium"
          color="black"
        >
          xstyled
        </x.h4>
        <x.p color="gray-500">A CSS-in-JS framework built for React.</x.p>
      </x.div>
    </x.div>
  )
}
```

## License

Licensed under the MIT License, Copyright © 2019-present Greg Bergé.

See [LICENSE](./LICENSE) for more information.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [3.1.2](https://github.com/gregberge/xstyled/compare/v3.1.1...v3.1.2) (2021-11-05)


### Bug Fixes

* **system:** fix caching issue with value with different types ([4a6ded3](https://github.com/gregberge/xstyled/commit/4a6ded3c89e2822bac4c672e06c2e74a8f2c7141))





## [3.1.1](https://github.com/gregberge/xstyled/compare/v3.1.0...v3.1.1) (2021-10-30)


### Bug Fixes

* allow numeric values for systemprops ([efdefe1](https://github.com/gregberge/xstyled/commit/efdefe1ddcf481a10bac1f7473143ce140e34ca8))
* boolean or string for transform property type ([1498d86](https://github.com/gregberge/xstyled/commit/1498d8671af5b46357ae99ae85c4ed62213e065f))
* color prop missing responsive support ([65f532b](https://github.com/gregberge/xstyled/commit/65f532bd1bf2f1b2dbc56068ef8d943223d7b4ff))
* ensure the number is one of the keys ([c2f71fd](https://github.com/gregberge/xstyled/commit/c2f71fd5c01b2f8597c4de370c39486218aa291b))
* more correct synthesized type that allows number keys ([c31c6b1](https://github.com/gregberge/xstyled/commit/c31c6b11b024fd9e5b96640f51d8f2b0c0342bb8))
* move typography color fix to proper file ([5c635b0](https://github.com/gregberge/xstyled/commit/5c635b08f321ce5c81ca7c2dbb93d29ca7ab9b12))
* transform prop type should be string not boolean ([50c847e](https://github.com/gregberge/xstyled/commit/50c847eddf11bcfb770830e98bdacda3514f4889))
* **types:** fix emotion createGlobalStyle type def ([95667bc](https://github.com/gregberge/xstyled/commit/95667bcb1d25ad008fd51080af66323b3be930ab))





# [3.1.0](https://github.com/gregberge/xstyled/compare/v3.0.3...v3.1.0) (2021-10-02)


### Bug Fixes

* **docs:** added a missing paren in the code example ([#303](https://github.com/gregberge/xstyled/issues/303)) ([901755b](https://github.com/gregberge/xstyled/commit/901755ba6d16eb621bacfe966886296cf2c6efdb))
* fix for Prototype Pollution ([#304](https://github.com/gregberge/xstyled/issues/304)) ([34ed147](https://github.com/gregberge/xstyled/commit/34ed147c2c9dd02bc213e74f9d5d7af1fae3b42c))
* **types:** fix wrong type for ThemeVariants ([#301](https://github.com/gregberge/xstyled/issues/301)) ([b69619e](https://github.com/gregberge/xstyled/commit/b69619ed06bedfe6146df35e697a5c60a79aa4d5))


### Features

* **emotion:** add importMap to the babel plugin ([#293](https://github.com/gregberge/xstyled/issues/293)) ([657f7fe](https://github.com/gregberge/xstyled/commit/657f7fe53b7f14a9523163aa0c1989655d83e058))
* **system:** add aria-disabled to the disabled state ([#297](https://github.com/gregberge/xstyled/issues/297)) ([d28cf38](https://github.com/gregberge/xstyled/commit/d28cf38425fef81c090cc95b47b62e2bfc440781))
* **typings:** assemble deep paths for autocompletion ([#310](https://github.com/gregberge/xstyled/issues/310)) ([9867158](https://github.com/gregberge/xstyled/commit/986715841c9c40c775d55ad54b8653165b2f1b71))





## [3.0.3](https://github.com/gregberge/xstyled/compare/v3.0.2...v3.0.3) (2021-06-25)


### Bug Fixes

* fix styled components module name ([#284](https://github.com/gregberge/xstyled/issues/284)) ([7c979d0](https://github.com/gregberge/xstyled/commit/7c979d0d69698629a516cfa5b8582cbc01aed1cb))





## [3.0.2](https://github.com/gregberge/xstyled/compare/v3.0.1...v3.0.2) (2021-06-18)


### Bug Fixes

* **breakpoints:** add _ into breakpoints ([bcfda4b](https://github.com/gregberge/xstyled/commit/bcfda4b77531487ac5a6b8c30b715dd2bcfd08bd)), closes [#277](https://github.com/gregberge/xstyled/issues/277)
* **emotion:** fix color prop ([4d88c18](https://github.com/gregberge/xstyled/commit/4d88c18712f9c020af988b8072157a836a6cd0d7)), closes [#276](https://github.com/gregberge/xstyled/issues/276)
* **system:** fix rpx transformers ([339f7fc](https://github.com/gregberge/xstyled/commit/339f7fcf59fb79b6e042564a03c07d40514a6c87)), closes [#274](https://github.com/gregberge/xstyled/issues/274)
* move away from mjs extension to work in more environments ([9098b83](https://github.com/gregberge/xstyled/commit/9098b83407888dea985081029dc93c18d5bb6eab))





# [2.5.0](https://github.com/gregberge/xstyled/compare/v2.4.1...v2.5.0) (2021-05-02)


### Bug Fixes

* **emotion:** add transient props support ([#252](https://github.com/gregberge/xstyled/issues/252)) ([ef34b61](https://github.com/gregberge/xstyled/commit/ef34b61282b126cb36c1ff92f874bc0af47baf5d))
* **system:** fix paddingBottom typings ([#242](https://github.com/gregberge/xstyled/issues/242)) ([ee01b9b](https://github.com/gregberge/xstyled/commit/ee01b9b42bf5a4de58227780bf1fccb41c489d51))
* **transform:** IE11 doesn't support dotAll ([#247](https://github.com/gregberge/xstyled/issues/247)) ([e620935](https://github.com/gregberge/xstyled/commit/e6209354c89739edd5205d15cd80de2a4afb2302))


### Features

* **style:** improve mixin ([#225](https://github.com/gregberge/xstyled/issues/225)) ([66ac294](https://github.com/gregberge/xstyled/commit/66ac294eb7fa0760bf75b1ac679e9e964397c915))





## [2.4.1](https://github.com/gregberge/xstyled/compare/v2.4.0...v2.4.1) (2021-03-29)


### Bug Fixes

* **styled:** fix nested states ([#221](https://github.com/gregberge/xstyled/issues/221)) ([78fac38](https://github.com/gregberge/xstyled/commit/78fac380e996ed93ca99020de668df2e52ccb318))





# [2.4.0](https://github.com/gregberge/xstyled/compare/v2.3.0...v2.4.0) (2021-03-28)


### Bug Fixes

* fix minW, minH, maxW, maxH ([#217](https://github.com/gregberge/xstyled/issues/217)) ([65c04de](https://github.com/gregberge/xstyled/commit/65c04deb8001835b4944fc9ac5dc34f2c7c64fab))


### Features

* add system.apply utility ([#218](https://github.com/gregberge/xstyled/issues/218)) ([c32f714](https://github.com/gregberge/xstyled/commit/c32f71489c506c095bc8fb75dc41add9be2fe300))

website/src/components/HomeHero.js
import React, { useState } from 'react'
import { x } from '@agriffis/xstyled-styled-components'
import { Highlight } from './Prism'

const codes = {
  x: `
<x.div
  p={{ _: 3, md: 6 }}
  bg="white"
  display="flex"
  spaceX={4}
  alignItems="center"
  boxShadow
  borderRadius
>
  <x.div flexShrink={0}>
    <img height={12} width={12} src="/logo.png" alt="xstyled" />
  </x.div>
  <x.div>
    <x.h4
      text={{ _: "md", lg: "xl" }}
      fontWeight="medium"
      color="black"
    >
      xstyled
    </x.h4>
    <x.p color="gray-500" text="sm" my={1}>
      A CSS-in-JS framework built for React.
    </x.p>
  </x.div>
</x.div>
`.trim(),
  styled: `
const Card = styled.div\`
  padding: 3 6;
  background-color: white;
  display: flex;
  align-items: center;
  box-shadow: default;
  border-radius: default;

  & > * {
    margin-right: 4;
    margin-left: 4;
  }
\`

const CardImage = styled.div\`
  flex-shrink: 0;
\`

const CardBody = styled.div\`\`

const CardTitle = styled.h4\`
  font-size: md;
  line-height: md;
  font-weight: medium;
  color: black;

  @media (min-width: lg) {
    font-size: lg;
    line-height: lg;
  } 
\`

const CardSubtitle = styled.p\`
  color: gray-500;
  font-size: sm;
  line-height: sm;
  margin-top: 1;
\`

<Card>
  <CardImage>
    <img height={12} width={12} src="/logo.png" alt="xstyled" />
  </CardImage>
  <CardBody>
    <CardTitle>xstyled</CardTitle>
    <CardSubtitle>
      A CSS-in-JS framework built for React.
    </CardSubtitle>
  </CardBody>
</Card>
  `.trim(),
}

const CodeButton = ({ active, ...props }) => (
  <x.button
    borderRadius="xl"
    bg={active ? 'fuchsia-500-a10' : 'transparent'}
    outline={{ focus: 0 }}
    ring={{ focusVisible: 2 }}
    ringColor="fuchsia-500-a50"
    border={0}
    fontWeight="semibold"
    transition
    color={active ? 'pink-300' : { _: 'white', hover: 'white-a80' }}
    px={2}
    py={1}
    {...props}
  />
)

function Code() {
  const [type, setType] = useState('x')
  const code = codes[type]
  return (
    <x.div
      position="relative"
      borderRadius
      w={{ lg: '90%' }}
      boxShadow="2xl"
      alignSelf={{ lg: 'flex-end' }}
      bg="cool-gray-800"
    >
      <x.div display="flex" bg="black-a25" spaceX={2} p={2}>
        <CodeButton active={type === 'x'} onClick={() => setType('x')}>
          x.*
        </CodeButton>
        <CodeButton
          active={type === 'styled'}
          onClick={() => setType('styled')}
        >
          styled.*
        </CodeButton>
      </x.div>
      <x.div overflow="auto" h={40 * 8}>
        <x.div position="relative" h="auto">
          <x.div
            hidden
            display={{ lg: 'block' }}
            position="absolute"
            top={0}
            bottom={0}
            left={0}
            bg="black-a25"
            w={9}
          />
          <x.pre display="flex" fontSize={{ _: 'xs', lg: 'sm' }}>
            <x.div
              aria-hidden="true"
              hidden
              display={{ lg: 'block' }}
              color="white-a50"
              py={4}
              pr={3}
              textAlign="right"
              userSelect="none"
              w={9}
            >
              {Array.from(
                { length: code.split('\n').length },
                (_, index) => index + 1,
              ).join('\n')}
            </x.div>
            <x.code
              flex={1}
              position="relative"
              color="white"
              display="block"
              p={4}
            >
              <Highlight language="jsx" code={code} />
            </x.code>
          </x.pre>
        </x.div>
      </x.div>
    </x.div>
  )
}

export function HomeHero() {
  return (
    <x.div
      py={{ _: 14, lg: 8 }}
      px={{ _: 2, lg: 5 }}
      mx={{ _: -3, lg: 0 }}
      w={{ _: '100vw', lg: 'auto' }}
      display="flex"
      flexDirection="column"
      gap={3}
      position="relative"
    >
      <x.div
        position="absolute"
        top={0}
        right={0}
        bottom={0}
        left={0}
        backgroundImage="gradient-to-r"
        gradientFrom="fuchsia-200"
        gradientTo="rose-200"
        boxShadow={{ _: null, lg: 'lg' }}
        transform
        rotate={{ lg: -6 }}
        borderRadius={{ lg: '3xl' }}
      />
      <x.div
        position="absolute"
        top={0}
        right={0}
        bottom={0}
        left={0}
        backgroundImage="gradient-to-r"
        gradientFrom="fuchsia-400"
        gradientTo="rose-500"
        boxShadow={{ _: null, lg: 'lg' }}
        transform
        rotate={{ lg: -2 }}
        borderRadius={{ lg: '3xl' }}
      />
      <x.div
        position="relative"
        p={{ _: 3, md: 6 }}
        w={{ lg: '70%' }}
        bg="white"
        display="flex"
        spaceX={4}
        alignItems="center"
        boxShadow
        borderRadius
      >
        <x.div flexShrink={0}>
          <x.img
            h={12}
            w={12}
            src="https://raw.githubusercontent.com/gregberge/xstyled/master/website/images/logo-manifest.png"
            alt="xstyled Logo"
          />
        </x.div>
        <x.div>
          <x.h4 text={{ _: 'md', lg: 'xl' }} fontWeight="medium" color="black">
            xstyled
          </x.h4>
          <x.p color="gray-500" text="sm" my={1}>
            A CSS-in-JS framework built for React.
          </x.p>
        </x.div>
      </x.div>
      <Code />
    </x.div>
  )
}
